todo_app/data/session_items.py

```diff
@@ -1,8 +1,75 @@
+from __future__ import annotations
+from typing import Callable, Dict, Optional
 from flask import session
+from dataclasses import dataclass
+
+
+@dataclass
+class Item:
+    status: str
+    title: str
+
+
+class Items:
+    _items_by_id: Dict[int, Item]
+    _next_item_id: int
+
+    def __init__(self, items_by_id: Dict[int, Item], next_item_id: int) -> None:
+        self._items_by_id = items_by_id
+        self._next_item_id = next_item_id
+
+    def items(self):
+        return self._items_by_id.items()
+
+    def default() -> Items:
+        ret = Items(items_by_id={}, next_item_id=0)
+        for item in _DEFAULT_ITEMS:
+            ret.add(item)
+        return ret
+
+    def add(self, item: Item):
+        next_id = self._next_item_id
+        self._next_item_id += 1
+        self._items_by_id[next_id] = item
+
+    def update(self, id: int, item: Item):
+        old_item = self.remove(id)
+        if old_item is None:
+            raise ValueError("Cannot update an item that does not exist!")
+        self._items_by_id[id] = item
+
+    def get(self, id: int) -> Optional[Item]:
+        return self._items_by_id.get(id)
+
+    def get_or_throw(self, id: int) -> Item:
+        item = self.get(id)
+        if item is None:
+            raise ValueError(
+                f"Could not find an item with id {id} - acceptable {self._items_by_id.keys()}"
+            )
+        return item
+
+    def remove(self, id: int) -> Optional[Item]:
+        return self._items_by_id.pop(id)
+
+    def save(self):
+        session["_items_by_id"] = self._items_by_id
+        session["_next_item_id"] = self._next_item_id
+
+    def load() -> Optional[Items]:
+        next_item_id = int(session.get("_next_item_id"))
+        if next_item_id is None:
+            return None
+        items_by_id_as_strings: Dict[str, Dict[str, str]] = session.get("_items_by_id")
```

Comment:
This code doesn't handle the case where we have no tasks - it causes the page to fail. I think the solution is to provide reasonable defaults, or maybe return early.